src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -257,86 +257,86 @@ public override async ValueTask<FormulaValue> Visit(CallNode node, EvalVisitorCo
+                }
+                else if (func is CustomTexlFunction customTexlFunc)
+                {
+                    // If custom function throws an exception, don't catch it - let it propagate up to the host.
+                    result = await customTexlFunc.InvokeAsync(FunctionServices, args, _cancellationToken).ConfigureAwait(false);
+                }
+                else if (func is UserDefinedFunction userDefinedFunc)
+                {
+                    UDFStackFrame frame = new UDFStackFrame(userDefinedFunc, args);
+                    UDFStackFrame framePop = null;
+
                     try
                     {
-                        result = await ptr(this, context.IncrementStackDepthCounter(childContext), node.IRContext, args).ConfigureAwait(false);
+                        _udfStack.Push(frame);
+
+                        (var irnode, _) = userDefinedFunc.GetIRTranslator();
+
+                        this.CheckCancel();
+
+                        result = await irnode.Accept(this, context).ConfigureAwait(false);
                     }
-                    catch (CustomFunctionErrorException ex)
+                    finally
                     {
-                        var irContext = node.IRContext;
-                        result = new ErrorValue(irContext, new ExpressionError() { Message = ex.Message, Span = irContext.SourceContext, Kind = ex.ErrorKind });
+                        framePop = _udfStack.Pop();
                     }
 
-                    if (!(result.IRContext.ResultType._type == node.IRContext.ResultType._type || result is ErrorValue || result.IRContext.ResultType is BlankType))
+                    if (frame != framePop)
                     {
-                        throw CommonExceptions.RuntimeMisMatch;
+                        throw new Exception("Something went wrong. UDF stack values didn't match.");
                     }
                 }
                 else
                 {
-                    result = CommonErrors.NotYetImplementedError(node.IRContext, $"Missing func: {func.Name}");
+                    if (FunctionImplementations.TryGetValue(func, out AsyncFunctionPtr ptr))
+                    {
+                        result = await ptr(this, context.IncrementStackDepthCounter(childContext), node.IRContext, args).ConfigureAwait(false);
+
+                        if (!(result.IRContext.ResultType._type == node.IRContext.ResultType._type || result is ErrorValue || result.IRContext.ResultType is BlankType))
+                        {
+                            throw CommonExceptions.RuntimeMisMatch;
+                        }
+                    }
+                    else
+                    {
+                        result = CommonErrors.NotYetImplementedError(node.IRContext, $"Missing func: {func.Name}");
+                    }
                 }
             }
+            catch (CustomFunctionErrorException ex)

```

Comment:
We need to catch this in all function dispatch cases so that IsError() can catch:

IsError(MyFunc())